src/core/common/time_ticker.hpp

```diff
@@ -0,0 +1,136 @@
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for a time ticker.
+ */
+
+#ifndef TIME_TICKER_HPP_
+#define TIME_TICKER_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "common/time.hpp"
+#include "common/timer.hpp"
+
+namespace ot {
+
+/**
+ * This class represents a time ticker.
+ *
+ * The time ticker emits periodic ticks (with 1 second period interval) to a set of registered tick receiver modules.
+ * The tick receivers (OpenThread objects) are identified by the `Receiver` enumeration. The receiver objects
+ * must provide `HandleTimeTick()` method which would be invoked by `TimeTicker` periodically.
+ *
+ */
+class TimeTicker : public InstanceLocator, private NonCopyable
+{
+public:
+    /**
+     * This enumeration represents time tick receivers.
+     *
+     * This enumeration contains the list of all OpenThread modules that can be registered as time tick receivers.
+     *
+     */
+    enum Receiver : uint8_t
+    {
+#if OPENTHREAD_FTD
+        kMleRouter,       ///< `Mle::MleRouter`
+        kAddressResolver, ///< `AddressResolver`
+#if OPENTHREAD_CONFIG_CHILD_SUPERVISION_ENABLE
+        kChildSupervisor, ///< `Utils::ChildSupervisor`
+#endif
+#endif
+#if OPENTHREAD_CONFIG_IP6_FRAGMENTATION_ENABLE
+        kIp6FragmentReassembler, ///< `Ip6::Ip6` (handling of fragmented messages)
+#endif
+#if OPENTHREAD_CONFIG_DUA_ENABLE || OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        kDuaManager, ///< `DuaManager`.
+#endif
+#if OPENTHREAD_CONFIG_MLR_ENABLE || OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+        kMlrManager, ///< `MlrManager`
+#endif
+        kMeshForwarder, ///< `MeshForwarder`
```

Comment:
Good idea. Changed it in new pushed commit.

I would like it if there was a way without adding a new `kNumReceivers` case in enum (since we use the `enum` as a type) but in this case, I agree it's better to define an explicit case in enum.